working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+at a high level in
+the
+[roadmap](https://github.com/dart-lang/language/blob/master/working/0110-incremental-sound-nnbd/roadmap.md).
+
+### Opted in libraries.
+
+Libraries and packages must opt into the feature as described elsewhere.  An
+opted-in library may depend on un-opted-in libraries, and vice versa.
+
+### Errors as warnings
+
+Weak null checking is enabled as soon as a package or library opts into this
+feature. When weak null checking is enabled, all errors specified this proposal
+(that is, all errors that arise only out of the new features of this proposal)
+shall be treated as warnings.
+
+Strong null checking is enabled by running the compilation or execution
+environment with the appropriate flags.  When strong null checking is enabled,
+errors specified in this proposal shall be treated as errors.
+
+### Legacy libraries
+
+Static checking for a library which has not opted into this feature (a *legacy*
+library) is done using the semantics as of the last version of the language
+before this feature ships (or the last version to which it has opted in, if that
+is different).  All opted-in libraries upstream from the legacy library are
+viewed by the legacy library with nullability related features erased from their
+APIs.  In particular:
+  - All types of the form `T?` in the opted-in API are treated as `T`.
+  - All required named parameters are treated as optional named parameters.
+  - The type `Never` is treated as the type `Null`
+
+In a legacy library, none of the new syntax introduced by this proposal is
+available, and it is a static error if it is used.
+
+### Importing legacy libraries from opted-in libraries
+
+The type system is extended with a notion of a legacy type operator.  For every
+type `T`, there is an additional type `T*` which is the legacy version of the
+type.  There is no surface syntax for legacy types, and implementations should
+display the legacy type `T*` in the same way that they would display the type
+`T`, except in so far as it is useful to communicate to programmers for the
+purposes of error messages that the type originates in unmigrated code.
+
+When static checking is done in a migrated library, types which are imported
+from unmigrated libraries are seen as legacy types.  However, for the purposes
+of type inference in migrated libraries, types imported from unmigrated
+libraries shall be treated as non-nullable.  As a result, legacy types will
+never appear as type annotations in migrated libraries, nor will they appear in
+reified positions.
+
+### Type reification
+
+All types reified in legacy libraries are reified as legacy types.  Runtime
+subtyping checks treat them according to the subtyping rules specified
+separately.
+
+### Runtime checks and weak checking
+
+When weak checking is enabled, runtime type tests (including explicit and
+implicit casts) shall succeed with a warning whenever the runtime type test
```

Comment:
Not sure what this means. What does "succeed with a warning mean", and what is the *result* of the test?
Does it succeed without a warning if the test succeeds with non-changed types?

This sounds like we have *runtime warnings* and each test needs to do two type checks (although we can perhaps merge that into one check algorithm with a ternary result). It sounds *expensive*. For the warning to be useful, it must at minimum remember the source position of the test, which is an extra overhead on application size.

Reply:
Yes to all of the above, except that it's not clear to me why you need to remember more information for the warning than you do for the error that would otherwise be thrown.